app/services/advance_order_service.rb

```diff
@@ -0,0 +1,35 @@
+class AdvanceOrderService
+  attr_reader :order
+
+  def initialize(order)
+    @order = order
+  end
+
+  def call(raise_on_error = false)
+    shipping_method_id = order.shipping_method.id if order.shipping_method.present?
+    options = { shipping_method_id: shipping_method_id }
+    raise_on_error ? advance_order!(options) : advance_order(options)
```

Comment:
I see this raise_on_error and advance_order! as YAGNI 🔥

Reply:
@luisramos0 This has implications in the handling in the places where this is used. For example, when creating a payment and the order cannot be transitioned all the way to completion, this would rollback. But in the edit order page, this would just stop where the state cannot be transitioned anymore.

I actually considered merging the two. But aside from raising an error where needed, I was being cautious about how `order.state == "complete"` is not the same as `order.completed?`. (An order that is completed and then cancelled would return `false` for the first condition but `true` for the next.)

I prefer not touch this for now, and think we should look into that later. What do you think? Spree 2.4-stable does not use the bang version in `admin/payments#create` anymore, and it can be :fire: without much audit then.

**Spree 2.4-stable:**

```
$ grep -r order.next | egrep -v 'spec|testing_support'
src/frontend/app/controllers/spree/orders_controller.rb:              @order.next if @order.cart?
src/frontend/app/controllers/spree/checkout_controller.rb:        unless @order.next
src/backend/app/controllers/spree/admin/orders/customer_details_controller.rb:            @order.next
src/backend/app/controllers/spree/admin/payments_controller.rb:            while @order.next; end
src/backend/app/controllers/spree/admin/payments_controller.rb:            # If "@order.next" didn't trigger payment processing already (e.g. if the order was
src/api/app/controllers/spree/api/checkouts_controller.rb:        @order.next!
src/api/app/controllers/spree/api/checkouts_controller.rb:        while @order.next; end
src/api/app/controllers/spree/api/checkouts_controller.rb:          if @order.completed? || @order.next
src/core/lib/tasks/exchanges.rake:        while order.state != order.checkout_steps[-2] && order.next; end
```

**Current Spree version:**

```
$ grep -r order.next | egrep -v 'spec|testing_support'
frontend/app/controllers/spree/orders_controller.rb:              @order.next_transition.run_callbacks if @order.cart?
frontend/app/controllers/spree/checkout_controller.rb:        unless @order.next
backend/app/controllers/spree/admin/orders/customer_details_controller.rb:            while @order.next; end
backend/app/controllers/spree/admin/orders_controller.rb:        while @order.next; end
backend/app/controllers/spree/admin/payments_controller.rb:              @order.next!
backend/app/controllers/spree/admin/payments_controller.rb:      # where order.next is called which leaves the order in payment step
api/app/controllers/spree/api/orders_controller.rb:        if @order.valid? && @order.next
api/app/controllers/spree/api/checkouts_controller.rb:          state_callback(:after) if @order.next
api/app/controllers/spree/api/checkouts_controller.rb:        @order.next!
api/app/controllers/spree/api/checkouts_controller.rb:          if @order.valid? && @order.next
```

Reply:
yeah, I see now, I think we need to keep it until 2.2

it was in v2.2 they removed this here: https://github.com/spree/spree/commit/4eade0fa78f426f5a3a8db15789dee48c2f87fa9?diff=unified
and then added the one without the bang here: https://github.com/spree/spree/commit/59198f74cf156e3b09dbf16f8eca523e28ebec79

Reply:
thanks for the explanation!

Reply:
The syntax is not clear though. You don't know what the boolean means from reading this:
```rb
AdvanceOrderService.new(@order).call(true)
```

You could go with a named parameter:
```rb
AdvanceOrderService.new(@order).call(raise_on_error: true)
```

But in this case I would go with the bang convention:
```rb
AdvanceOrderService.new(@order).call!
```

If you put `options` into its own method it becomes really easy.

Reply:
Thanks, @mkllnk! I changed this here: https://github.com/openfoodfoundation/openfoodnetwork/pull/3560/commits/94606025c9c881c61fe52c07dc9adf4f86f35962 I also added a couple of specs to cover `call!`.